k-distribution/src/main/scripts/lib/pyk/pyk/subst.py

```diff
@@ -0,0 +1,113 @@
+from dataclasses import dataclass
+from itertools import chain
+from typing import Dict, Iterator, Mapping, Optional, TypeVar
+
+from .kast import KApply, KInner, KRewrite, KSequence, KToken, KVariable
+from .kastManip import bottom_up
+from .utils import FrozenDict
+
+K = TypeVar('K')
+V = TypeVar('V')
+
+
+@dataclass(frozen=True)
+class Subst(Mapping[str, KInner]):
+    subst: FrozenDict[str, KInner]
+
+    def __init__(self, subst: Mapping[str, KInner] = {}):
+        object.__setattr__(self, 'subst', FrozenDict({k: v for k, v in subst.items() if type(v) is not KVariable or v.name != k}))
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self.subst)
+
+    def __len__(self) -> int:
+        return len(self.subst)
+
+    def __getitem__(self, key: str) -> KInner:
+        return self.subst[key]
+
+    def __mul__(self, other: 'Subst') -> 'Subst':
+        return self.compose(other)
+
+    def __call__(self, term: KInner) -> KInner:
+        return self.apply(term)
+
+    def compose(self, other: 'Subst') -> 'Subst':
+        from_other = ((k, self(v)) for k, v in other.items())
+        from_self = ((k, v) for k, v in self.items() if k not in other)
+        return Subst(dict(chain(from_other, from_self)))
```

Comment:
So in `subst1.compose(subst2)`, we are preferring the bindings from `subst2`? The name `compose` seems ambiguous about which is preferred, but `update` would be pretty clear I think.

Reply:
Semantically, `Subst` can be interpreted as a total function over variables:

* if `(x, v) in subst.items()` for some `v` then semantically `subst(x) = v`;
* otherwise, `susbt(x) = x`.

Assuming this interpretation, this method implements function composition:

* For all `(x, v) in from_other`, we have `v = subst1(subst2(x))`. Moreover, `(x, v) in from_other` for some `v` iff `x in susbt2`.
* For all `(x, v) in from_self`, we have `v = subst1(x)`. Moreover, `(x, v) in from_self` for some `v` iff `x in susbt1` and `x not in susbt2`. Thus we have `v = subst1(subst2(x))`.
* For all `x` such that `(x, v) not in from_other` and `(x, v) not in from_self` for any `v`, we have `x = susbt1(subst2(x))`.

Reply:
Nice, good intuition here then :)